chain/errors.go

```diff
@@ -0,0 +1,495 @@
+
+	case ErrMaxFeeExceeded:
+		return "max fee exceeded"
+
+	case ErrTxAlreadyKnown:
+		return "txn already known"
+
+	case ErrTxAlreadyConfirmed:
+		return "Transaction already in block chain"
+
+	case ErrMempoolConflict:
+		return "txn mempool conflict"
+
+	case ErrReplacementAddsUnconfirmed:
+		return "replacement adds unconfirmed"
+
+	case ErrInsufficientFee:
+		return "insufficient fee"
+
+	case ErrTooManyReplacements:
+		return "too many potential replacements"
+
+	case ErrMempoolMinFeeNotMet:
+		return "mempool min fee not met"
+
+	case ErrConflictingTx:
+		return "bad txns spends conflicting tx"
+
+	case ErrEmptyOutput:
+		return "bad txns vout empty"
+
+	case ErrEmptyInput:
+		return "bad txns vin empty"
+
+	case ErrTxTooSmall:
+		return "tx size small"
+
+	case ErrDuplicateInput:
+		return "bad txns inputs duplicate"
+
+	case ErrEmptyPrevOut:
+		return "bad txns prevout null"
+
+	case ErrBelowOutValue:
+		return "bad txns in belowout"
+
+	case ErrNegativeOutput:
+		return "bad txns vout negative"
+
+	case ErrLargeOutput:
+		return "bad txns vout toolarge"
+
+	case ErrLargeTotalOutput:
+		return "bad txns txouttotal toolarge"
+
+	case ErrScriptVerifyFlag:
+		return "mandatory script verify flag failed"
+
+	case ErrTooManySigOps:
+		return "bad txns too many sigops"
+
+	case ErrInvalidOpcode:
+		return "disabled opcode"
+
+	case ErrTxAlreadyInMempool:
+		return "txn already in mempool"
+
+	case ErrMissingInputs:
+		return "missing inputs"
+
+	case ErrOversizeTx:
+		return "bad txns oversize"
+
+	case ErrCoinbaseTx:
+		return "coinbase"
+
+	case ErrNonStandardVersion:
+		return "version"
+
+	case ErrNonStandardScript:
+		return "scriptpubkey"
+
+	case ErrBareMultiSig:
+		return "bare multisig"
+
+	case ErrScriptSigNotPushOnly:
+		return "scriptsig not pushonly"
+
+	case ErrScriptSigSize:
+		return "scriptsig size"
+
+	case ErrTxTooLarge:
+		return "tx size"
+
+	case ErrDust:
+		return "dust"
+
+	case ErrMultiOpReturn:
+		return "multi op return"
+
+	case ErrNonFinal:
+		return "non final"
+
+	case ErrNonBIP68Final:
+		return "non BIP68 final"
+
+	case ErrSameNonWitnessData:
+		return "txn same nonwitness data in mempool"
+
+	case ErrNonMandatoryScriptVerifyFlag:
+		return "non mandatory script verify flag"
+	}
+
+	return "unknown error"
+}
+
+// BtcdErrMap takes the errors returned from btcd's `testmempoolaccept` and
+// `sendrawtransaction` RPCs and map them to the errors defined above, which
+// are results from calling either `testmempoolaccept` or `sendrawtransaction`
+// in `bitcoind`.
+//
+// Errors not mapped in `btcd`:
+//   - deployment error from `validateSegWitDeployment`.
+//   - the error when total inputs is higher than max allowed value from
+//     `CheckTransactionInputs`.
+//   - the error when total outputs is higher than total inputs from
+//     `CheckTransactionInputs`.
+//   - errors from `CalcSequenceLock`.
+//
+// NOTE: This is not an exhaustive list of errors, but it covers the
```

Comment:
Hmm I am still a bit surprised why we are moving the error definition to the btcwallet pkg, imo they fit better into the btcd pkg, especially the btcd errors should be defined there anyways because its the official client?

Reply:
IMO it's better here. First of all, this set of newly defined error types are not `btcd` errors, they are RPC errors, returned by either calling `sendrawtransaction` or in the `reject-reason` field in a response returned by `testmempoolaccept`. When in `lnd`, we don't care about the specific backend, rather, we just want to get a defined error and act on it.

This means whether it's `btcd`, `bitcoind`, or `neutrino`, we should have a place to catch the implementation-specific errors, and "translate" them into a uniformized RPC error. `chain` is a good place to perform this task as it acts like a backend-multiplexer - it provides a standardized interface `chain.Interface`, and handles the implementation-specific logic here.

On the other hand, `btcd` should just be, well, `btcd`, an independent implementation, that's oblivious about how `neutrino` or `bitcoind` handles stuff, which is why in addition to this error handling, I also think the `btcd/rpcclient` should be moved here. The overall design is, that `btcwallet` provides multi-backend handling, and `btcd` just being a protocol implementation.

Reply:
Great explanation thank you